File: code/Algorithm/leetcode/2020.9-(30)/96.不同的二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=96 lang=cpp
 *
 * [96] 不同的二叉搜索树
 *
 * https://leetcode-cn.com/problems/unique-binary-search-trees/description/
 *
 * algorithms
 * Medium (69.07%)
 * Likes:    812
 * Dislikes: 0
 * Total Accepted:    85.7K
 * Total Submissions: 124K
 * Testcase Example:  '3'
 *
 * 给定一个整数 n，求以 1 ... n 为节点组成的二叉搜索树有多少种？
 * 
 * 示例:
 * 
 * 输入: 3
 * 输出: 5
 * 解释:
 * 给定 n = 3, 一共有 5 种不同结构的二叉搜索树:
 * 
 * ⁠  1         3     3      2      1
 * ⁠   \       /     /      / \      \
 * ⁠    3     2     1      1   3      2
 * ⁠   /     /       \                 \
 * ⁠  2     1         2                 3
 * 
 */

// catalan公式 记住就好

// @lc code=start
class Solution
{
public:
    int numTrees(int n)
    {
        int dp[n + 1];
        dp[0] = 1;
        dp[1] = 1;

        for (int i = 2; i <= n; i++)
        {
            int curSum = 0;
            for (int j = 0; j < i; j++)
                curSum += dp[j] * dp[i - j - 1];
            dp[i] = curSum;
        }

        return dp[n];
    }
};
// @lc code=end

File: code/JavaLab/src/main/java/algorithm/leetcode/tree_50/96.不同的二叉搜索树.java
package algorithm.leetcode.tree_50;

/**
 * https://leetcode-cn.com/problems/unique-binary-search-trees/
 * <p>
 * 给你一个整数 n ，求恰由 n 个节点组成且节点值从 1 到 n 互不相同的 二叉搜索树 有多少种？返回满足题意的二叉搜索树的种数。
 */
class LC96 {


    public int numTrees(int n) {

        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1; // 一个节点只有一种情况


        for (int i = 2; i <= n; i++) {

            for (int j = 0; j < i; j++) {
                // 左子树种类情况
                int leftNum = dp[j];
                // 右子树种类情况(去掉根节点)
                int rightNum = dp[i - j - 1];
                // 左右子树情况做笛卡尔积即可
                dp[i] += leftNum * rightNum;
            }
        }
        return dp[n];
    }
}